File: src/jk_5/nailed/teams/Team.java
package jk_5.nailed.teams;

import com.google.common.collect.Lists;
import jk_5.nailed.Nailed;
import jk_5.nailed.map.Map;
import jk_5.nailed.players.Player;
import jk_5.nailed.util.EnumColor;
import net.minecraft.src.ChunkCoordinates;
import net.minecraft.src.ScorePlayerTeam;

import java.util.List;

/**
 * No description given
 *
 * @author jk-5
 */
public class Team {

    public static Team UNKNOWN = new Team("Unknown Team", "unknown", EnumColor.GREY);

    private String name;
    private String teamId;
    private EnumColor color;
    private Player leader = null;
    private boolean ready = false;
    private boolean friendlyFire = false;
    public ScorePlayerTeam scoreboardTeam = null;

    private Map map;
    private ChunkCoordinates spawn = null;

    public Team(String name, String teamId, EnumColor color) {
        this.name = name;
        this.teamId = teamId;
        this.color = color;
    }

    public EnumColor getColor() {
        return this.color;
    }

    public String getName() {
        return this.name;
    }

    public String getTeamID() {
        return this.teamId;
    }

    public void sendChatMessage(String message) {
        for (Player p : Nailed.playerRegistry.getPlayers()) {
            if (p.getTeam() == this) {
                p.sendChatMessage(message);
            }
        }
    }

    @Override
    public String toString() {
        return String.format("%s%s%s", this.color.toString(), this.name, EnumColor.RESET);
    }

    public String toChatFormatString() {
        if (this == Team.UNKNOWN) return "";
        return String.format("%s* %s%s ", this.getColor().toString(), this.getName(), EnumColor.RESET);
    }

    public void setMap(Map map){
        this.map = map;
    }

    public void setSpawnpoint(ChunkCoordinates coordinates){
        this.spawn = coordinates;
    }

    public boolean shouldOverrideSpawnpoint(){
        if(this.map == null) return false;
        if(this.spawn == null) return false;
        return true;
    }

    public ChunkCoordinates getSpawnpoint(){
        if(!this.shouldOverrideSpawnpoint()) return null;
        else return this.spawn;
    }

    public Player getTeamLeader(){
        return this.leader;
    }

    public void setTeamLeader(Player player){
        this.leader = player;
    }

    public List<String> getAllPlayerNames(){
        List<String> ret = Lists.newArrayList();
        for(Player p : Nailed.playerRegistry.getPlayers()){
            if(p.getTeam() == this) ret.add(p.getUsername());
        }
        return ret;
    }

    public boolean isReady(){
        return this.ready;
    }

    public void setReady(boolean ready){
        this.ready = ready;
        if(this.isReady()){
            this.map.sendMessageToAllPlayers("Team " + this.toString() + " is ready!");
        }else{
            this.map.sendMessageToAllPlayers("Team " + this.toString() + " is not ready!");
        }
        this.map.getGameThread().notifyReadyUpdate();
    }

    public boolean friendlyFireEnabled(){
        return this.friendlyFire;
    }

    public void setFriendlyFireEnabled(boolean on){
        this.friendlyFire = on;
    }
}

File: src/main/scala/jk_5/nailed/team/Team.scala
package jk_5.nailed.team

import jk_5.nailed.util.EnumColor
import jk_5.nailed.players.{PlayerRegistry, Player}
import jk_5.nailed.map.Map
import net.minecraft.scoreboard.ScorePlayerTeam
import net.minecraft.util.ChunkCoordinates

/**
 * No description given
 *
 * @author jk-5
 */
object Team{
  final val UNKNOWN = new Team("unknown")
}
case class Team(private final val teamId: String) {
  private var name: String = this.teamId
  private var color: EnumColor = _
  private var leader: Player = _
  private var ready = false
  private var friendlyFire = false
  private var scoreboardTeam: ScorePlayerTeam = _
  private var map: Map = _
  private var spawn: ChunkCoordinates = _

  @inline def sendChatMessage(message: String) = PlayerRegistry.getPlayers.filter(_.getTeam == this).foreach(_.sendChatMessage(message))
  @inline override def toString = this.color.toString + this.name + EnumColor.RESET
  @inline def shouldOverrideSpawnpoint = if(this.map == null) false else if(this.spawn == null) false else true
  @inline def toChatFormatString = this match{
    case Team.UNKNOWN => ""
    case t => "%s* %s%s ".format(this.getColor.toString, this.getName, EnumColor.RESET)
  }

  def setReady(ready: Boolean){
    this.ready = ready
    if(this.isReady) this.map.sendMessageToAllPlayers("Team " + this.toString + " is ready!")
    else this.map.sendMessageToAllPlayers("Team " + this.toString + " is ready!")
    this.map.getGameThread.notifyReadyUpdate()
  }

  @inline def getColor = this.color
  @inline def getName = this.name
  @inline def getTeamID = this.teamId
  @inline def getSpawnpoint = if(this.shouldOverrideSpawnpoint) this.spawn else null
  @inline def getTeamLeader = this.leader
  @inline def isReady = this.ready
  @inline def friendlyFireEnabled = this.friendlyFire
  @inline def getScoreboardTeam: Option[ScorePlayerTeam] = Option(this.scoreboardTeam)
  @inline def getAllPlayerNames = PlayerRegistry.getPlayers.filter(_.getTeam == this).map(_.getUsername)

  @inline def setMap(map: Map) = this.map = map
  @inline def setSpawnpoint(coords: ChunkCoordinates) = this.spawn = coords
  @inline def setTeamLeader(player: Player) = this.leader = player
  @inline def setFriendlyFireEnabled(on: Boolean) = this.friendlyFire = on
  @inline def setName(name: String) = this.name = name
  @inline def setColor(color: EnumColor) = this.color = color
  @inline def setScoreboardTeam(st: ScorePlayerTeam) = this.scoreboardTeam = st
}
